src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -330,69 +389,130 @@ new rfprec maxcsize = MCompactIndex
-    lastMinRfbits <- readSTRef mciLastMinRfbits
-    let lb = smaybe NoBound (\i -> Bound (fromIntegral i) Exclusive) lastMinRfbits
-        ub = Bound (fromIntegral minRfbits) Inclusive
-        x  = fromIntegral pageNr
-    writeRange mciRangeFinder lb ub x
-    writeSTRef mciLastMinRfbits $! SJust minRfbits
-
-    -- Fill primary vector
-    readSTRef mciPrimary >>= \vum -> VUM.write vum ix minPrimbits
+    goAppend pageNr ix
+  where
+    (minKey, maxKey) = minMax app
 
-    -- Fill clash vector and tie-breaker map
-    lastMaxPrimbits <- readSTRef mciLastMaxPrimbits
-    let clash = lastMaxPrimbits == SJust minPrimbits
-    readSTRef mciClashes >>= \vum -> VUM.write vum ix (Bit clash)
-    when clash $ modifySTRef' mciTieBreaker (IntMap.insert pageNr minKey)
-    writeSTRef mciLastMaxPrimbits $! SJust maxPrimbits
+    minRfbits :: Word16
+    minRfbits = topBits16 mciRangeFinderPrecision minKey
 
-    pure res
-  where
-      minRfbits :: Word16
-      minRfbits = topBits16 mciRangeFinderPrecision minKey
+    minPrimbits, maxPrimbits :: Word32
+    minPrimbits = sliceBits32 mciRangeFinderPrecision minKey
+    maxPrimbits = sliceBits32 mciRangeFinderPrecision maxKey
 
-      minPrimbits, maxPrimbits :: Word32
-      minPrimbits = sliceBits32 mciRangeFinderPrecision minKey
-      maxPrimbits = sliceBits32 mciRangeFinderPrecision maxKey
+    -- | Yield a chunk and start a new one if the current chunk is already full.
+    --
+    -- TODO(optimisation): yield will eagerly allocate new mutable vectors, but
+    -- maybe that should be done lazily.
+    yield :: ST s (Maybe Chunk)
+    yield = do
+        pageNr <- readSTRef mciCurrentPageNumber
+        if pageNr `mod` mciMaxChunkSize == 0 then do -- The current chunk is full
+          cPrimary <- VU.unsafeFreeze =<< readSTRef mciPrimary
+          (writeSTRef mciPrimary $!) =<< VUM.new mciMaxChunkSize
+          cClashes <- VU.unsafeFreeze =<< readSTRef mciClashes
+          (writeSTRef mciClashes $!) =<< VUM.new mciMaxChunkSize
+          pure $ Just (Chunk{..})
+        else -- the current chunk is not yet full
+          pure Nothing
+
+    -- | Fill primary and clash vectors for a larger-than-page value. Yields
+    -- chunks if necessary
+    overflows :: Int -> ST s [Chunk]
+    overflows incr
+      | incr <= 0 = pure []
+      | otherwise = do
+          pageNr <- readSTRef mciCurrentPageNumber
+          let ix = pageNr `mod` mciMaxChunkSize -- will be 0 in recursive calls
+              remInChunk = min incr (mciMaxChunkSize - ix)
+          readSTRef mciPrimary >>= \vum ->
+            writeRange vum (BoundInclusive ix) (BoundExclusive $ ix + remInChunk) minPrimbits
+          readSTRef mciClashes >>= \vum ->
+            writeRange vum (BoundInclusive ix) (BoundExclusive $ ix + remInChunk) (Bit True)
+          writeSTRef mciCurrentPageNumber $! pageNr + remInChunk
+          res <- yield
+          maybe id (:) res <$> overflows (incr - remInChunk)
+
+    -- | Meat of the function
```

Comment:
For readability, perhaps we could put this one before all the phases it calls, rather than in the middle.